completion_aggregator/aggregator.py

```diff
@@ -22,22 +23,41 @@
 
 EnrollmentTuple = collections.namedtuple('EnrollmentTuple', ['username', 'course_key'])
 
-MAX_KEYS_PER_TASK = 128
+MAX_KEYS_PER_TASK = 16
 
-
-def perform_aggregation():
+def perform_aggregation(batch_size=None, limit=None, delay=0.0, routing_key=None):
     """
     Enqueues tasks to reaggregate modified completions.
 
     When blocks are completed, they mark themselves as stale.  This function
     collects all stale blocks for each enrollment, and enqueues a single
     recalculation of all aggregators containing those stale blocks.
+
+    batch_size (int|None):
+        Maximum number of stale completions to enqueue in a single run.
+        None = Process all outstanding StaleCompletions
+
+    delay (float):
+        The amount of time to wait between batches.
+
+    routing_key (str|None):
+        A routing key to pass to celery for the update_aggregators tasks.
     """
-    with transaction.atomic():
-        stale_queryset = models.StaleCompletion.objects.filter(resolved=False)
+    stale_queryset = models.StaleCompletion.objects.filter(resolved=False)
+    initial_count = stale_queryset.count()
+    if limit is None:
+        limit = float('inf')
+    limit = min(limit, initial_count)
+    enqueued = 0
+    task_options = {}
+    if routing_key:
+        task_options['routing_key'] = routing_key
+    while enqueued < limit:
         stale_blocks = collections.defaultdict(set)
         forced_updates = set()
-        for stale in stale_queryset:
+        evaluated = stale_queryset[:batch_size]  # This actually works fine if batch_size is None
```

Comment:
If `batch_size` is say 100, isn't this going to keep processing the same 100 entries over and over? I don't see anything that forces the query to restart.

Also, if some other process handled a few StaleCompletions while this ran, is it possible this could become an infinite loop, because `len(evaluated)` could be 0 and `enqueued` will never reach `limit`?

Reply:
Yup.  Updated to use a pk-based indexing scheme.